### ISSUE TITLE
(Windows) High Memory/RAM Usage

### ISSUE THREAD

**Issue by DigitalNXZ** [Nheko-Reborn/nheko#898]
### Describe the bug

I think there is a memory leakage~?
or is this normal?

Edit: also I found out when click in a room on the header (room settings - pops up and also the memory usage goes up by about 30MB and even if I close the settings it stays and goes even higher~ also when I kicked someone same problem appeared?);

### To Reproduce

~idk just had it open without doing anything for 'bout 40 minutes (idling)

### What happened?

Had Nheko open for 'bout 40 Minutes in the background, and out of nowhere it jumped up to 800MB and stabilized at 500MB~ (and than it freezes at 500MB)
Tried: Restart, and reinstall

(removed a few things from the logs.. i'm new to matrix so idk if you can do weird stuff with the weird looking codes int the logs)

### Expected behavior

at least when I open it, it always has 100-150MB, shouldn't be still kinda be under 200-300MB?

### Screenshots

![image](https://user-images.githubusercontent.com/53005759/149703792-a77ebc9f-0ce8-4679-a93a-bf14876a8cf3.png)

### Version

0.9.1-dfab0c92 | Windows

### Operating system

Windows

### Installation method

Windows download

### Did you use profiles?

- [ ] Profiles used?

### Relevant log output

```shell
[2022-01-17 02:52:29.931] [ui] [info] Restoring window size 0x0
[2022-01-17 02:52:30.529] [ui] [info] Loaded jdenticon plugin.
[2022-01-17 02:52:35.214] [ui] [info] starting nheko 0.9.1-dfab0c92
[2022-01-17 02:53:13.783] [net] [error] Autodiscovery failed. Unknown error when requesting .well-known. 0 7
[2022-01-17 02:53:20.552] [net] [info] Login requested to user server: https://matrix-client.matrix.org/
[2022-01-17 02:53:20.676] [db] [info] initializing LMDB
[2022-01-17 02:53:20.681] [db] [info] database ready
[2022-01-17 02:53:20.682] [crypto] [info] creating new olm account
[2022-01-17 02:53:20.683] [crypto] [info] Online key backup disabled.
[2022-01-17 02:53:20.683] [crypto] [info] ed25519   : (removed by me, just in case..)
[2022-01-17 02:53:20.683] [crypto] [info] curve25519: (removed by me, just in case..)
[2022-01-17 02:53:20.683] [crypto] [info] generating one time keys
[2022-01-17 02:53:20.716] [db] [info] Storing secret 'pickle_secret' successful
[2022-01-17 02:53:20.846] [net] [info] uploaded 50 signed_curve25519 one-time keys
[2022-01-17 02:53:20.846] [net] [info] trying initial sync
[2022-01-17 02:53:21.320] [net] [info] initial sync completed
[2022-01-17 02:53:21.323] [db] [info] mark room (removed by me, just in case..) as encrypted
[2022-01-17 02:53:21.466] [db] [info] Invalidating self verification status
[2022-01-17 02:53:22.758] [net] [info] TURN server(s) retrieved from homeserver:
[2022-01-17 02:53:22.758] [net] [info] username: (removed by me, just in case..)
[2022-01-17 02:53:22.758] [net] [info] ttl: 86400 seconds
[2022-01-17 02:53:22.758] [net] [info] uri: turn:turn.matrix.org:3478?transport=udp
[2022-01-17 02:53:22.758] [net] [info] uri: turn:turn.matrix.org:3478?transport=tcp
[2022-01-17 02:53:22.758] [net] [info] uri: turns:turn.matrix.org:443?transport=tcp
[2022-01-17 02:53:25.999] [db] [info] Clicked verify master key with passphrase
[2022-01-17 02:53:31.751] [db] [info] Storing secret 'm.cross_signing.master' successful
[2022-01-17 02:53:31.777] [db] [info] Storing secret 'm.cross_signing.self_signing' successful
[2022-01-17 02:53:31.801] [db] [info] Storing secret 'm.cross_signing.user_signing' successful
[2022-01-17 02:53:31.827] [db] [info] Storing secret 'm.megolm_backup.v1' successful
[2022-01-17 02:53:31.882] [db] [info] Invalidating self verification status
[2022-01-17 02:53:31.934] [db] [info] Invalidating self verification status
[2022-01-17 02:54:41.799] [ui] [warning] Already loading older messages
[2022-01-17 02:54:42.430] [ui] [warning] Already loading older messages
[2022-01-17 02:54:43.105] [ui] [warning] Already loading older messages
[2022-01-17 02:55:02.939] [net] [error] sync error: (parser: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected end of input; expected '[', '{', or a literal: )
[2022-01-17 02:55:10.493] [ui] [info] Restoring window size 958x1048
[2022-01-17 02:55:10.855] [ui] [info] Loaded jdenticon plugin.
[2022-01-17 02:55:11.856] [ui] [info] starting nheko 0.9.1-dfab0c92
[2022-01-17 02:55:12.101] [db] [info] database ready
[2022-01-17 02:55:12.101] [db] [info] restoring state from cache
[2022-01-17 02:55:12.101] [db] [info] Invalidating self verification status
[2022-01-17 02:55:12.104] [db] [info] Restored 4 rooms from cache
[2022-01-17 02:55:12.140] [crypto] [info] ed25519   : (removed by me, just in case..)
[2022-01-17 02:55:12.140] [crypto] [info] curve25519: (removed by me, just in case..)
[2022-01-17 02:55:12.140] [crypto] [info] Online key backup disabled.
[2022-01-17 02:55:12.256] [crypto] [info] Fetched server key count 50 signed_curve25519
[2022-01-17 02:55:12.304] [db] [info] Invalidating self verification status
[2022-01-17 02:55:14.297] [net] [info] TURN server(s) retrieved from homeserver:
[2022-01-17 02:55:14.297] [net] [info] username: (removed by me, just in case..)
[2022-01-17 02:55:14.297] [net] [info] ttl: 86400 seconds
[2022-01-17 02:55:14.297] [net] [info] uri: turn:turn.matrix.org:3478?transport=udp
[2022-01-17 02:55:14.297] [net] [info] uri: turn:turn.matrix.org:3478?transport=tcp
[2022-01-17 02:55:14.297] [net] [info] uri: turns:turn.matrix.org:443?transport=tcp
[2022-01-17 02:57:42.480] [net] [error] sync error: (parser: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected end of input; expected '[', '{', or a literal: )
[2022-01-17 02:57:51.025] [ui] [info] Restoring window size 1920x1017
[2022-01-17 02:57:51.397] [ui] [info] Loaded jdenticon plugin.
[2022-01-17 02:57:52.264] [ui] [info] starting nheko 0.9.1-dfab0c92
[2022-01-17 02:57:52.631] [db] [info] database ready
[2022-01-17 02:57:52.631] [db] [info] restoring state from cache
[2022-01-17 02:57:52.631] [db] [info] Invalidating self verification status
[2022-01-17 02:57:52.634] [db] [info] Restored 4 rooms from cache
[2022-01-17 02:57:52.650] [crypto] [info] ed25519   : (removed by me, just in case..)
[2022-01-17 02:57:52.650] [crypto] [info] curve25519: (removed by me, just in case..)
[2022-01-17 02:57:52.650] [crypto] [info] Online key backup disabled.
[2022-01-17 02:57:52.768] [crypto] [info] Fetched server key count 50 signed_curve25519
[2022-01-17 02:57:52.789] [db] [info] Invalidating self verification status
[2022-01-17 02:57:54.692] [net] [info] TURN server(s) retrieved from homeserver:
[2022-01-17 02:57:54.692] [net] [info] username: (removed by me, just in case..)
[2022-01-17 02:57:54.692] [net] [info] ttl: 86400 seconds
[2022-01-17 02:57:54.692] [net] [info] uri: turn:turn.matrix.org:3478?transport=udp
[2022-01-17 02:57:54.692] [net] [info] uri: turn:turn.matrix.org:3478?transport=tcp
[2022-01-17 02:57:54.692] [net] [info] uri: turns:turn.matrix.org:443?transport=tcp
[2022-01-17 02:58:20.783] [net] [error] sync error: (parser: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected end of input; expected '[', '{', or a literal: )
```

**Comment by deepbluev7** [comment]
Nheko caches about 100MB of messages and images. But what you are seeing is probably just mapping of the DB. The actual memory usage is usually not calculated properly in the various task managers, because Nheko mmaps the whole db into memory. As such if you have free RAM, the memory usage will look higher than it is, when you are close to running out of memory (since then the OS just unmaps the corresponding memory pages, if they weren't accessed recently). If you run Nheko under massif or so, real memory usage should show ~200MB.

**Comment by deepbluev7** [comment]
See also https://lmdb.readthedocs.io/en/release/#memory-usage

**Comment by DigitalNXZ** [comment]
> Nheko caches about 100MB of messages and images. But what you are seeing is probably just mapping of the DB. The actual memory usage is usually not calculated properly in the various task managers, because Nheko mmaps the whole db into memory. As such if you have free RAM, the memory usage will look higher than it is, when you are close to running out of memory (since then the OS just unmaps the corresponding memory pages, if they weren't accessed recently). If you run Nheko under massif or so, real memory usage should show ~200MB.

oh~ aright? well I had it open today in the morning and it had about 1.2~GB, is the DB that massive? Since I don't really have chats or messages, just a room that I created to test "Matrix"~ is there not a simpler solution? at the end of the day it takes way to much :,D (even if its not Electron) its a problem when I play games.. up to 300MB would be fine enough

How can I run it under massif? I would like to try it out ^^

**Comment by deepbluev7** [comment]
Sorry, I have no idea how to use massif on Windows :D 

And Nheko shouldn't take that much, but I am not familiar with Windows memory management. I'd test it under actual memory pressure, if it doesn't become much less in that case.

**Comment by DigitalNXZ** [comment]
> Sorry, I have no idea how to use massif on Windows :D
> 
> And Nheko shouldn't take that much, but I am not familiar with Windows memory management. I'd test it under actual memory pressure, if it doesn't become much less in that case.

could this help? or is there something I can try out for ya?
[capture.csv](https://github.com/Nheko-Reborn/nheko/files/7883612/capture.csv)
(I captured nheko over "Very Sleepy")

**Comment by deepbluev7** [comment]
Hm, not sure, if I read that correctly, but it looks like a lot of images are allocated and never freed? Can you maybe try our nightly build and see, if that behaves differently? (It's linked in the README)

Ftr, on Linux memory usage looks like this over a few hours (It was idle in the long flat part, I only interacted at the start and end with it.):
![image](https://user-images.githubusercontent.com/16312324/149827596-1c9c7145-737c-4795-89b7-d442e832454e.png)

**Comment by DigitalNXZ** [comment]
> Hm, not sure, if I read that correctly, but it looks like a lot of images are allocated and never freed? Can you maybe try our nightly build and see, if that behaves differently? (It's linked in the README)
> 
> Ftr, on Linux memory usage looks like this over a few hours (It was idle in the long flat part, I only interacted at the start and end with it.): ![image](https://user-images.githubusercontent.com/16312324/149827596-1c9c7145-737c-4795-89b7-d442e832454e.png)

Tested Nightly (0.9.1-429d51d4)
https://www.youtube.com/watch?v=gQvV9MLtDa8

Still have it open since the end of the reocrding.. it sits on 780MB rn
Everyone I kick from rooms adds a few MB, every setting I open adds quite a few MB, texts don't seem to really affect it.

**Comment by deepbluev7** [comment]
I think we have seen that before, but I could never reproduce it... #838 for example... I'll look into it.

**Comment by redsky17** [comment]
I haven't seen this issue and I regularly run nheko on windows.  For example, I've been running nheko for at least 24 hours and using it here and there throughout the day and my RAM usage is right around 150MB:

![image](https://user-images.githubusercontent.com/962117/150231682-5c23dabb-ab26-425c-8915-95b7aaab99c8.png)

I'll certainly keep an eye out for it.

**Comment by DigitalNXZ** [comment]
> I haven't seen this issue and I regularly run nheko on windows. For example, I've been running nheko for at least 24 hours and using it here and there throughout the day and my RAM usage is right around 150MB:
> 
> ![image](https://user-images.githubusercontent.com/962117/150231682-5c23dabb-ab26-425c-8915-95b7aaab99c8.png)
> 
> Even after scrolling the timeline for several days passed dozens of images, it only goes up by a little bit:
> 
> ![image](https://user-images.githubusercontent.com/962117/150231783-fcca2c62-8961-44ff-a4d6-d0072da19dc1.png)
> 
> (after this point, it goes back down again).
> 
> I'll certainly keep an eye out for it.

Hi! ^^ weird, for me nothing goes back, it just goes up and stays there even after hours~
maybe a problem for specific versions? or just for Windows 11?
or my cpu, gpu well i don't thinks anything of this could be the problem.. may the ram?

**Comment by redsky17** [comment]
I'm also on windows 11.  I can't rule out a windows 11-specific bug or anything, but it's at least not universal...  I doubt it's a hardware issue on your end as well.

**Comment by DigitalNXZ** [comment]
> I'm also on windows 11. I can't rule out a windows 11-specific bug or anything, but it's at least not universal... I doubt it's a hardware issue on your end as well.

hi, also
in the last day I did a whole pc reset, it didn't fix anything
now I updated multiple drivers/hardware? and it seems like may the CPU was the Culprit? Same day before the update I was around 500mb without goin' down, well depends, if u clicked on way more buttons and so on it was even higher (version 0.9.1-9f31095e), after the the update it jumps to 350mb idling.. still kinda high but ya 

CPU: Ryzen 5 1600

A Question, why does the usage spikes up so drastically when clicking on the channel headers? every time? does it don't cache it somehow? likely 20mb with every click is a bit.. :o 

Since my English is not the best i'll explain it again this time on german since I saw some german words in @deepbluev7 Screenshot~

--------

Hi! Also, in den Letzen Tage habe ich meinen PC zurueckgesetzt (Werkeinstellung) einfach mal so.. um einfach alles muell was sich die Jahre angesammelt hat loszuwerden.

Weiter geholfen im Bereich High Memory Usage hat es bei Nehko nicht.. 

Dann habe ich ein paar Updates ausgefuehrt die ich vorher nicht hatte, da AMD ja selbst meint Updates nicht ausfuehren wenn man keine Probleme mit der jetzigen hat zmdst bei deren Ryzen 5.. was auch immer.. Habe ich jetzt einfach mal gemacht und es scheint zum teil geholfen zu haben.. 

Habe es vor und nach dem Update getestet
Davor 500mb, im Idling ging's weiterhin auch nicht runter, evtl hoeher jenachdem wie viele man gekickt hatte ausm raum und allgemein knoepfe gedrueckt hat..
Nachdem Update 500mb, geht aber runter im Idling auf ca 350mb. quite high but irgendwie scheint ein teil davon jetzt gefixt zu sein?

CPU: Ryzen 5 1600


Aber mal so eine nebenfrage.. wieso steigt der usage eigtl so krass mit jedem click auf den channel header? und immer aufs neue? 20mb sind schon ziemlich hoch? :0
